In [1]:
import pandas as pd 
import numpy as np

In [2]:
transaction_data = pd.read_csv("data/transaction_data.csv")
product = pd.read_csv("data/product.csv")
hh_demographic = pd.read_csv("data/hh_demographic.csv")
campaign_table = pd.read_csv("data/campaign_table.csv")
campaign_desc = pd.read_csv("data/campaign_desc.csv")
coupon = pd.read_csv("data/coupon.csv")
coupon_redempt = pd.read_csv("data/coupon_redempt.csv")

In [3]:
transaction_data.shape

(2595732, 12)

In [4]:
transaction_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 2595732 entries, 0 to 2595731
Data columns (total 12 columns):
 #   Column             Dtype  
---  ------             -----  
 0   household_key      int64  
 1   BASKET_ID          int64  
 2   DAY                int64  
 3   PRODUCT_ID         int64  
 4   QUANTITY           int64  
 5   SALES_VALUE        float64
 6   STORE_ID           int64  
 7   RETAIL_DISC        float64
 8   TRANS_TIME         int64  
 9   WEEK_NO            int64  
 10  COUPON_DISC        float64
 11  COUPON_MATCH_DISC  float64
dtypes: float64(4), int64(8)
memory usage: 237.6 MB


In [5]:
transaction_data.isnull().sum()

household_key        0
BASKET_ID            0
DAY                  0
PRODUCT_ID           0
QUANTITY             0
SALES_VALUE          0
STORE_ID             0
RETAIL_DISC          0
TRANS_TIME           0
WEEK_NO              0
COUPON_DISC          0
COUPON_MATCH_DISC    0
dtype: int64

transaction_data.duplicated().sum()

La tabla transacciones no tiene valores nulos ni filas duplicadas en sus 
2,595,732 registros. Los datos llegan completos a nivel de transacción, 
por lo que no se requiere imputación ni deduplicación en esta tabla.

In [6]:
transaction_data.describe()

,household_key,BASKET_ID,DAY,PRODUCT_ID,QUANTITY,SALES_VALUE,STORE_ID,RETAIL_DISC,TRANS_TIME,WEEK_NO,COUPON_DISC,COUPON_MATCH_DISC
count,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06
mean,1.271953e+03,3.402620e+10,3.887562e+02,2.891435e+06,1.004286e+02,3.104120e+00,3.142673e+03,-5.387054e-01,1.561586e+03,5.622150e+01,-1.641600e-02,-2.918564e-03
std,7.260660e+02,4.711649e+09,1.897210e+02,3.837404e+06,1.153436e+03,4.182274e+00,8.937113e+03,1.249191e+00,3.998378e+02,2.710223e+01,2.168410e-01,3.969004e-02
min,1.000000e+00,2.698485e+10,1.000000e+00,2.567100e+04,0.000000e+00,0.000000e+00,1.000000e+00,-1.800000e+02,0.000000e+00,1.000000e+00,-5.593000e+01,-7.700000e+00
25%,6.560000e+02,3.040805e+10,2.290000e+02,9.174590e+05,1.000000e+00,1.290000e+00,3.300000e+02,-6.900000e-01,1.308000e+03,3.300000e+01,0.000000e+00,0.000000e+00
50%,1.272000e+03,3.276081e+10,3.900000e+02,1.028816e+06,1.000000e+00,2.000000e+00,3.720000e+02,-1.000000e-02,1.613000e+03,5.600000e+01,0.000000e+00,0.000000e+00
75%,1.913000e+03,4.012685e+10,5.530000e+02,1.133018e+06,1.000000e+00,3.490000e+00,4.220000e+02,0.000000e+00,1.843000e+03,8.000000e+01,0.000000e+00,0.000000e+00
max,2.500000e+03,4.230536e+10,7.110000e+02,1.831630e+07,8.963800e+04,8.400000e+02,3.428000e+04,3.990000e+00,2.359000e+03,1.020000e+02,0.000000e+00,0.000000e+00


In [7]:
(transaction_data['QUANTITY'] < 0).sum()

np.int64(0)

In [8]:
(transaction_data['SALES_VALUE'] < 0).sum()  

np.int64(0)

In [9]:
(transaction_data['RETAIL_DISC'] < 0).sum()   

np.int64(1303026)

QUANTITY y SALES_VALUE no presentan valores negativos, pero RETAIL_DISC sí: 
1,303,026 transacciones (≈50% del total) tienen un descuento aplicado, 
registrado con signo negativo respecto al precio de lista. SALES_VALUE 
representa el precio pagado por el cliente, no el precio de lista original.

In [11]:
hh_demographic.shape

(801, 8)

In [12]:
transaction_data['household_key'].nunique()

2500

solo 801 de los 2,500 hogares (32%) tienen datos demográficos disponibles 
en hh_demographic. cualquier análisis por edad, ingreso o tamaño de hogar 
solo puede reportarse sobre este subconjunto del 32%, y no debe extrapolarse 
al resto de la base de clientes.

In [13]:
product.info()

<class 'pandas.DataFrame'>
RangeIndex: 92353 entries, 0 to 92352
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   PRODUCT_ID            92353 non-null  int64
 1   MANUFACTURER          92353 non-null  int64
 2   DEPARTMENT            92353 non-null  str  
 3   BRAND                 92353 non-null  str  
 4   COMMODITY_DESC        92353 non-null  str  
 5   SUB_COMMODITY_DESC    92353 non-null  str  
 6   CURR_SIZE_OF_PRODUCT  92353 non-null  str  
dtypes: int64(2), str(5)
memory usage: 4.9 MB


In [14]:
product.isnull().sum()

PRODUCT_ID              0
MANUFACTURER            0
DEPARTMENT              0
BRAND                   0
COMMODITY_DESC          0
SUB_COMMODITY_DESC      0
CURR_SIZE_OF_PRODUCT    0
dtype: int64

In [15]:
product.duplicated().sum()

np.int64(0)

In [16]:
product['DEPARTMENT'].nunique()

44

In [17]:
product['DEPARTMENT'].value_counts()

DEPARTMENT
GROCERY            39021
DRUG GM            31529
PRODUCE             3118
COSMETICS           3011
NUTRITION           2914
MEAT                2544
MEAT-PCKGD          2427
DELI                2354
PASTRY              2149
FLORAL               938
SEAFOOD-PCKGD        563
MISC. TRANS.         490
SPIRITS              377
SEAFOOD              369
GARDEN CENTER        128
RESTAURANT           102
MISC SALES TRAN       88
SALAD BAR             48
COUP/STR & MFG        39
TRAVEL & LEISUR       28
FROZEN GROCERY        23
KIOSK-GAS             16
                      15
CHEF SHOPPE           14
RX                     9
CNTRL/STORE SUP        4
GM MERCH EXP           3
TOYS                   3
DAIRY DELI             3
VIDEO RENTAL           3
POSTAL CENTER          3
GRO BAKERY             2
CHARITABLE CONT        2
PROD-WHS SALES         2
DELI/SNACK BAR         2
AUTOMOTIVE             2
PHOTO                  2
VIDEO                  2
ELECT &PLUMBING        1
MEAT-WHSE     

In [18]:
product[product['DEPARTMENT'] == '']

,PRODUCT_ID,MANUFACTURER,DEPARTMENT,BRAND,COMMODITY_DESC,SUB_COMMODITY_DESC,CURR_SIZE_OF_PRODUCT


In [19]:
(product['DEPARTMENT'] == '').sum()

np.int64(0)

In [20]:
hh_demographic.info()

<class 'pandas.DataFrame'>
RangeIndex: 801 entries, 0 to 800
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   classification_1   801 non-null    str  
 1   classification_2   801 non-null    str  
 2   classification_3   801 non-null    str  
 3   HOMEOWNER_DESC     801 non-null    str  
 4   classification_5   801 non-null    str  
 5   classification_4   801 non-null    str  
 6   KID_CATEGORY_DESC  801 non-null    str  
 7   household_key      801 non-null    int64
dtypes: int64(1), str(7)
memory usage: 50.2 KB


In [21]:
hh_demographic.isnull().sum()

classification_1     0
classification_2     0
classification_3     0
HOMEOWNER_DESC       0
classification_5     0
classification_4     0
KID_CATEGORY_DESC    0
household_key        0
dtype: int64

In [22]:
hh_demographic.duplicated().sum()

np.int64(0)

In [23]:
hh_demographic['classification_1'].unique()

<StringArray>
['Age Group6', 'Age Group4', 'Age Group2', 'Age Group3', 'Age Group1',
 'Age Group5']
Length: 6, dtype: str

In [24]:
print(hh_demographic['classification_2'].unique())
print(hh_demographic['classification_3'].unique())
print(hh_demographic['classification_4'].unique())
print(hh_demographic['classification_5'].unique())
print(hh_demographic['HOMEOWNER_DESC'].unique())
print(hh_demographic['KID_CATEGORY_DESC'].unique())

<StringArray>
['X', 'Y', 'Z']
Length: 3, dtype: str
<StringArray>
[ 'Level4',  'Level5',  'Level3',  'Level6',  'Level1',  'Level7',  'Level2',
  'Level8',  'Level9', 'Level12', 'Level10', 'Level11']
Length: 12, dtype: str
<StringArray>
['2', '3', '4', '1', '5+']
Length: 5, dtype: str
<StringArray>
['Group5', 'Group4', 'Group3', 'Group6', 'Group2', 'Group1']
Length: 6, dtype: str
<StringArray>
['Homeowner', 'Unknown', 'Renter', 'Probable Renter', 'Probable Owner']
Length: 5, dtype: str
<StringArray>
['None/Unknown', '1', '2', '3+']
Length: 4, dtype: str


In [28]:
hh_demographic = hh_demographic.rename(columns={
    'classification_1': 'AGE_DESC',
    'classification_2': 'MARITAL_STATUS',
    'classification_3': 'INCOME_DESC',
    'classification_4': 'HOUSEHOLD_SIZE_DESC',
    'classification_5': 'HH_COMP_DESC'
})

In [31]:
hh_demographic.columns

Index(['AGE_DESC', 'MARITAL_STATUS', 'INCOME_DESC', 'HOMEOWNER_DESC',
       'HH_COMP_DESC', 'HOUSEHOLD_SIZE_DESC', 'KID_CATEGORY_DESC',
       'household_key'],
      dtype='str')

In [32]:
hh_demographic.columns

Index(['AGE_DESC', 'MARITAL_STATUS', 'INCOME_DESC', 'HOMEOWNER_DESC',
       'HH_COMP_DESC', 'HOUSEHOLD_SIZE_DESC', 'KID_CATEGORY_DESC',
       'household_key'],
      dtype='str')

nota: las columnas de hh_demographic venían anonimizadas como classification_1 
a classification_5. se identificó su significado comparando los valores 
contra la documentación del dataset original y se renombraron a AGE_DESC, 
MARITAL_STATUS, INCOME_DESC, HOUSEHOLD_SIZE_DESC y HH_COMP_DESC.

In [33]:
campaign_table.info()

<class 'pandas.DataFrame'>
RangeIndex: 7208 entries, 0 to 7207
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   DESCRIPTION    7208 non-null   str  
 1   household_key  7208 non-null   int64
 2   CAMPAIGN       7208 non-null   int64
dtypes: int64(2), str(1)
memory usage: 169.1 KB


In [34]:
campaign_table.isnull().sum()

DESCRIPTION      0
household_key    0
CAMPAIGN         0
dtype: int64

In [35]:
campaign_table.duplicated().sum()

np.int64(0)

In [36]:
campaign_table['DESCRIPTION'].unique()

<StringArray>
['TypeA', 'TypeB', 'TypeC']
Length: 3, dtype: str

In [37]:
campaign_desc.info()

<class 'pandas.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   DESCRIPTION  30 non-null     str  
 1   CAMPAIGN     30 non-null     int64
 2   START_DAY    30 non-null     int64
 3   END_DAY      30 non-null     int64
dtypes: int64(3), str(1)
memory usage: 1.1 KB


In [38]:
campaign_desc.isnull().sum()

DESCRIPTION    0
CAMPAIGN       0
START_DAY      0
END_DAY        0
dtype: int64

In [39]:
campaign_desc.duplicated().sum()

np.int64(0)

In [40]:
coupon.info()

<class 'pandas.DataFrame'>
RangeIndex: 124548 entries, 0 to 124547
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype
---  ------      --------------   -----
 0   COUPON_UPC  124548 non-null  int64
 1   PRODUCT_ID  124548 non-null  int64
 2   CAMPAIGN    124548 non-null  int64
dtypes: int64(3)
memory usage: 2.9 MB


In [41]:
coupon.isnull().sum()

COUPON_UPC    0
PRODUCT_ID    0
CAMPAIGN      0
dtype: int64

In [42]:
coupon.duplicated().sum()

np.int64(5164)

In [43]:
coupon['COUPON_UPC'].nunique()

1135

In [44]:
coupon[coupon.duplicated(keep=False)].head(10)

,COUPON_UPC,PRODUCT_ID,CAMPAIGN
47,54100010032,34214,27
49,54100010032,34214,27
50,54100010032,34214,27
51,54100010032,34214,27
124,54589314175,48728,27
126,54589314175,48728,27
127,54589314187,48728,27
128,54589314187,48728,27
129,54589314187,48728,27
130,54589314187,48728,27


nota: coupon tiene 5,164 filas exactamente duplicadas (mismo COUPON_UPC, 
PRODUCT_ID y CAMPAIGN repetidos), no solo la relación 1:m esperada. antes 
de mergear contra transaction_data o product, se debe deduplicar con 
coupon.drop_duplicates() para evitar inflar los resultados.

In [45]:
coupon_redempt.info()

<class 'pandas.DataFrame'>
RangeIndex: 2318 entries, 0 to 2317
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   household_key  2318 non-null   int64
 1   DAY            2318 non-null   int64
 2   COUPON_UPC     2318 non-null   int64
 3   CAMPAIGN       2318 non-null   int64
dtypes: int64(4)
memory usage: 72.6 KB
